### PR TITLE
Painless: Modify Loader to Load Classes Directly from Definition

### PR DESCRIPTION
**Pull request by jdconrad**
Updated the Painless Loader to first look for classes that are already loaded as part of the Definition initialization through the Whitelist.  This allows for modules/plugins to whitelist custom classes without forcing Painless to have a direct relationship with the module's/plugin's ClassLoader.

This also removes the singleton of Definition, so now there will be one Definition per Compiler.  However, the standard whitelist files singleton still exists out of convenience for now for testing.  This should be cleaned up later in a different PR.

@rjernst This will be difficult to properly test until we integrate directly with SPI.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine Please test this.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java [hunk 1]
```diff
@@ -101,9 +101,11 @@ public PainlessScriptEngine(Settings settings, Collection<ScriptContext<?>> cont
 
         for (ScriptContext<?> context : contexts) {
             if (context.instanceClazz.equals(SearchScript.class) || context.instanceClazz.equals(ExecutableScript.class)) {
-                contextsToCompilers.put(context, new Compiler(GenericElasticsearchScript.class, Definition.DEFINITION));
+                contextsToCompilers.put(context, new Compiler(GenericElasticsearchScript.class, new Definition(
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: For now, let's move the construction of Definition out of the loop, so we still only create one, until we can figure out if we need an internal cache to ensure too many copies aren't loaded of common classes.

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: Agreed.  Changed this to re-use the same Definition for all contexts for now.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review.  Will push as soon as the CI completes successfully.

**Pull request merged**